### DISCUSSION TITLE
Migratable VMs on MooseFS

### DISCUSSION THREAD

**Discussion by kcperry** [moosefs/moosefs#578] [Q&A]
I'm coming [back] to MooseFS after using LizardFS for a good number of years.  One of the things that I had an issue with was being able to live migrate VMs (which I do as  files, not NBDs).  I see talk about libvirt but I do not use that and as far as I can tell, it should not be necessary.  Just following along with [](https://www.linux-kvm.org/page/Migration), I have attempted live migration on MFS but what I ultimately get in the QEMU console after the memory is transferred is a message about the VM file not being available.  It is the same message I would get on LizardFS as well.

For the record, shutting down and restarting a VM works as expected but it would be great to have the ability to live migrate.

My question is, is this known to work on MFS and if so, where are some published use cases.  If not, is it on the road map  for the community edition of Moose 4.

**Comment by xandrus** [comment]
Can you share with us the command line you use for live migration?
Perhaps you will be able to share some some logs with us?

The live migration works without any issues with libvirt, Proxmox and XCP-ng, so I suspect this is more of some kind of misconfiguration.

**Comment by kcperry** [comment]
My test was this...

I started a VM on one (source) server with:

qemu-system-x86_64 -enable-kvm -net nic -net user -daemonize -m 2048 -vnc localhost:6 -usb -usbdevice tablet -cdrom ubuntu-22.04.1-desktop-amd64.iso

on the target system, I then did,

qemu-system-x86_64 -enable-kvm -net nic -net user -daemonize -m 2048 -vnc localhost:6 -usb -usbdevice tablet -cdrom ubuntu-22.04.1-desktop-amd64.iso -incoming tcp:0:4444

The VNC console for this system will display "Guest has not initialized the display (yet)" until I initiate the migration in the next step.

Going into the QEMU console for the source server I do:
migrate -d tcp:<target server IP>:4444  ("info migrate" will show the memory being transferred)

Once I see the VNC console on the target server go live, I kill the VM on the source server.

Next , I go back back to the target server and into it's QEMU console.  This time I do see the status as running (so I must have botched a step last time) and the "info migrate" shows complete but the VM is frozen.  If I do a stop and cont, I don't get an error (but it is still frozen- see the attached image).

Feels like I'm closer now. 

![image](https://github.com/moosefs/moosefs/assets/24230758/7108a66a-a10f-4f62-b71d-57dfddae93de)

**Comment by xandrus** [comment]
By the way. What is the version of qemu?

`qemu-system-x86_64 --version`

**Comment by kcperry** [comment]
Ahh, I forgot to mention that.  The above test was done on 5.1 which is very old so I built 9.0.1 with this configuration:

_configure --enable-kvm --enable-sdl --enable-vnc --enable-spice --enable-tools --enable-linux-aio --enable-slirp --enable-fuse_

Also, both systems are Ubuntu 22.04 LTS with both the fuse2 and fuse3 libraries.  When I check the binaries with ldd they reference the fuse3 objects.

Now when I attempt a migration, I ultimately get a migration status failed... Something about qemu-savevm_state_complete_precopy_non_iterable and bdrv_inactivate_all.  The target VM also shuts down.  I built these with the fuse option because I was under the impression that that was needed but but perhaps it is not?  Clearly something is off with my build.

Here's the image...
![image](https://github.com/user-attachments/assets/9a9cec49-44c9-4aa6-a470-b7d6323364bd)

**Comment by xandrus** [comment]
Hi,
thank you for all the information.

I would like to add that I did the QEMU live migration without any problems.
This is how my test configuration looks like:
2x QEMU machines:
- Ubuntu: 22.04
- QEMU: QEMU emulator version 6.2.0 (Debian 1:6.2+dfsg-2ubuntu6.21)
- MooseFS client 3.0.116(Ubuntu build)
- Both machines have a MooseFS cluster mounted with standard parameters:
`mfsmount -H mfsmater.my.lan /mnt/mfs`

My step by step guide:
1. Create 8GB disk image for operating system
`qemu-img create ubuntu.qcow2 8G -f qcow2`
2. Start VM and install OS via VNC
```
qemu-system-x86_64 -monitor stdio -enable-kvm -net nic -net user -m 2048 -display none -vnc :0,password=on -usb -usbdevice tablet -cdrom ubuntu-20.04.iso -drive file=/mnt/mfs/qemu/ubuntu.qcow2,media=disk,format=qcow2

#inside qemu console set password for VNC:
(qemu) change vnc password
```
3. After the operating system installation is complete, update the packages and install qemu-guest-agent
`apt install qemu-guest-agent`
4.  Power off machine and start it without cdrom on source machine.
`qemu-system-x86_64 -monitor stdio -enable-kvm -net nic -net user -m 2048 -display none -vnc :0,password=on -usb -usbdevice tablet -drive file=/mnt/mfs/qemu/ubuntu.qcow2,media=disk,format=qcow2`
5. On target machine start QEMU with parameters:
```
qemu-system-x86_64 -monitor stdio -enable-kvm -net nic -net user -m 2048 -display none -vnc :0,password=on -usb -usbdevice tablet -drive file=/mnt/mfs/qemu/ubuntu.qcow2,media=disk,format=qcow2 -incoming tcp:0:4444

#inside qemu console set password for VNC:
(qemu) change vnc password
```
6. On the source machine, run the migration and check the status
```
migrate -d tcp:[TARGET IP]:4444

(qemu) info migrate
globals:
store-global-state: on
only-migratable: off
send-configuration: on
send-section-footer: on
decompress-error-check: on
clear-bitmap-shift: 18
Migration status: completed
total time: 5316 ms
downtime: 11 ms
setup: 0 ms
transferred ram: 708842 kbytes
throughput: 1092.83 mbps
remaining ram: 0 kbytes
total ram: 2114632 kbytes
duplicate: 353919 pages
skipped: 0 pages
normal: 176089 pages
normal bytes: 704356 kbytes
dirty sync count: 4
page size: 4 kbytes
multifd bytes: 0 kbytes
pages-per-second: 32710
```
7.  After logging into the target VNC, everything worked flawlessly, and during the migration the "ping 8.8.8.8" operation was performed, which worked fine.

I would also like to add that I have repeated the migration process many times and have not been able to get any operating system freezes.
Check if your installed operating system has qemu-guest-agent on board, it may help a lot.
Note that I added the `-display none` parameter to the qemu command to not display the video output at all.

**Comment by kcperry** [comment]
Ok, some updates...  TL;DR this still isn't working for me but I've made some progress based on your use case.  That is to say, I did a clean install of 22.04 and  also installed qemu-guest-agent via package manager.  Here's where I am now:

1) Compile string correction... In addition to what I have above, it appears "--enable-guest-agent" is also needed
2) Launch string...  

`qemu-system-x86_64 -enable-kvm -cpu EPYC -net nic,model=virtio,macaddr=aa:bb:cc:23:44:62 -net tap,ifname=test,script=no -m 2048 -monitor stdio -display none -usb -usbdevice tablet -drive file=test-vm-migration.qcow2,media=disk,if=virtio -chardev socket,path=/tmp/qga.sock,server=on,wait=off,id=qga0 -device virtio-serial -device virtserialport,chardev=qga0,name=org.qemu.guest_agent.0`

My launch string needed to include "**-chardev socket,path=/tmp/qga.sock,server=on,wait=off,id=qga0 -device virtio-serial -device virtserialport,chardev=qga0,name=org.qemu.guest_agent.0**" otherwise the the guest agent would not start.

3) Necessity of "**-monitor stdio -display none**"...  this makes no sense to me.  When I don't include this I get the exact same issue as before (https://github.com/moosefs/moosefs/discussions/578#discussioncomment-10034058).  When I do include it, I  get a successful migration;

```
(qemu) info migrate
globals:
store-global-state: on
only-migratable: off
send-configuration: on
send-section-footer: on
decompress-error-check: on
clear-bitmap-shift: 18
Migration status: completed
total time: 75914 ms
downtime: 349 ms
setup: 1 ms
transferred ram: 1148721 kbytes
throughput: 123.96 mbps
remaining ram: 0 kbytes
total ram: 2114376 kbytes
duplicate: 276684 pages
skipped: 0 pages
normal: 285947 pages
normal bytes: 1143788 kbytes
dirty sync count: 6
page size: 4 kbytes
multifd bytes: 0 kbytes
pages-per-second: 4993
precopy ram: 1144752 kbytes
downtime ram: 3707 kbytes
```

but the target system crashes with;

```
QEMU 9.0.1 monitor - type 'help' for more information
(qemu) qemu-system-x86_64: Failed to put registers after init: Invalid argument
qemu-system-x86_64: network script /usr/local/bin/../etc/qemu-ifdown failed with status 256
```

The script error should be benign since I don't use any up/down scripts when initializing.  Also, this is occurring whether I specify the CPU or not.   I once read that that was necessary to do but in your case you did not specify it so it must be related to something else.   I'm still digging through bug reports but unless you have any other ideas, I'm going to engage with the QEMU folks since this really isn't a MooseFS issue.  Thank you for all your time thus far.
